tests/test_databases.py

```diff
@@ -543,6 +543,33 @@ async def test_transaction_rollback(database_url):
             assert len(results) == 0
 
 
+@pytest.mark.parametrize("database_url", DATABASE_URLS)
+@async_adapter
+async def test_transaction_doesnt_leak_when_commit_fails(database_url):
+    """
+    Ensure that transaction doesn't leak the connection when the commit or rollback
+    fails
+    """
+
+    async with Database(database_url) as database:
+        with pytest.raises(Exception) as excinfo:
+            async with database.connection() as connection:
+                await connection.execute(
+                    """
+                    CREATE TABLE test (id integer PRIMARY KEY INITIALLY DEFERRED);
+                    """
+                )
+                async with connection.transaction():
+                    await connection.execute("insert into test (id) values (1)")
+                    await connection.execute("insert into test (id) values (1)")
+
+        # During transaction.commit() postgres will raise this exception:
+        #  asyncpg.exceptions.UniqueViolationError: duplicate key value violates unique constraint "test_pkey"
+        #  DETAIL:  Key (id)=(1) already exists.
+        assert "unique constraint" in str(excinfo.value)
+        assert connection._connection_counter == 0
```

Comment:
@aminalaee I think this test it's only valid for postgresql. Is there a way to skip for other databases?

Reply:
Yeah sure, https://github.com/encode/databases/blob/6a52a4a37e69fcb25f309ff5d38e4d43eb80b32c/tests/test_databases.py#L482